crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -1969,33 +1969,75 @@ impl<'db> TypeInferenceBuilder<'db> {
         } = assignment;
 
         for target in targets {
-            self.infer_target(target, value);
+            self.infer_target(target, value, |_, ty| ty);
         }
     }
 
-    /// Infer the definition types involved in a `target` expression.
+    /// Infer the (definition) types involved in a `target` expression.
     ///
     /// This is used for assignment statements, for statements, etc. with a single or multiple
-    /// targets (unpacking).
+    /// targets (unpacking). If `target` is an attribute expression, we check that the assignment
+    /// is valid. For 'target's that are definitions, this check happens elsewhere.
+    ///
+    /// The `to_assigned_ty` function is used to convert the inferred type of the `value` expression
+    /// to the type that is eventually assigned to the `target`.
     ///
     /// # Panics
     ///
     /// If the `value` is not a standalone expression.
-    fn infer_target(&mut self, target: &ast::Expr, value: &ast::Expr) {
+    fn infer_target<F>(&mut self, target: &ast::Expr, value: &ast::Expr, to_assigned_ty: F)
+    where
+        F: Fn(&'db dyn Db, Type<'db>) -> Type<'db>,
```

Comment:
I went through three or four different iterations of this API here. I'm still not satisfied. If we want to pass in the `assigned_ty` directly, we need to call `infer_standalone_expression` at the call site. But we are only allowed to do that if the `target` is not a `ast::Expr::Name(_)`, so we need to basically copy the whole `match` statement below to the call site. The duplication maybe not so bad, but I could some expression types are not yet handled here (see pre-existing TODO), and then we would need to remember to update it everywhere.

Maybe my struggle to incorporate this here is also a sign that I'm doing something wrong?

Reply:
I think it's fine. An altenrative, to avoid monomorphization is to have a `Target` enum with a `Name`, and another variant but naming kind of gets awkward.

Reply:
> What's the reason that we only infer the value's type for name? Is it because the inference for non-names happens elsewhere and it, therefore, avoids doubl-inference?

We only call `infer_standalone_expression(value)` for *non*-`Name`s. For `Name`s, we call `infer_definition`, which (somewhere down the line) calls `infer_standalone_expression(value)` itself.

So yes, if we would call `infer_standalone_expression(value)` for all targets, we would get double-inference (results e.g. in duplicated diagnostics).